core/templates/components/button-directives/create-activity-button.component.spec.ts

```diff
@@ -0,0 +1,285 @@
+  }));
+
+  it('should initialize and call createNewExploration()', fakeAsync(() => {
+    const userServiceSpy = spyOn(userService, 'getUserInfoAsync')
+      // This throws "Argument of type 'Promise<{ isLoggedIn: () => boolean;
+      // canCreateCollections: () => boolean; }>' is not assignable to parameter
+      // of type 'Promise<UserInfo>'.". This is because the actual
+      // 'getUserInfoAsync' returns more properties than required. We need to
+      // suppress this error because we need only "isLoggedIn" and
+      // "canCreateCollections" functions for testing.
+      // @ts-expect-error
+      .and.returnValue(Promise.resolve({
+        isLoggedIn: () => true,
+        canCreateCollections: () => false
+      }));
+    const urlParamsSpy = spyOn(urlService, 'getUrlParams').and.returnValue({
+      mode: 'create'
+    });
+    const explorationCreationServiceSpy = spyOn(
+      explorationCreationService, 'createNewExploration');
+
+    expect(component.canCreateCollections).toBe(false);
+    expect(component.userIsLoggedIn).toBe(false);
+
+    component.ngOnInit();
+    tick();
+    fixture.detectChanges();
+
+    expect(userServiceSpy).toHaveBeenCalled();
+    expect(urlParamsSpy).toHaveBeenCalled();
+    expect(explorationCreationServiceSpy).toHaveBeenCalled();
+    expect(component.canCreateCollections).toBe(false);
+    expect(component.userIsLoggedIn).toBe(true);
+  }));
+
+  it('should return from initCreateProcess() when' +
+    ' creation is in progress', () => {
+    component.creationInProgress = true;
+    expect(component.initCreationProcess()).toBe();
+  });
+
+  it('should create new exploration if user cannot create collections', () => {
+    component.creationInProgress = false;
+    component.canCreateCollections = false;
+    const explorationCreationServiceSpy = spyOn(
+      explorationCreationService, 'createNewExploration');
+
+    component.initCreationProcess();
+
+    expect(explorationCreationServiceSpy).toHaveBeenCalled();
+  });
+
+  it('should replace location', () => {
+    component.creationInProgress = false;
+    component.canCreateCollections = true;
+    const urlServiceSpy = spyOn(urlService, 'getPathname').and.returnValue(
+      'not/creator-dashboard');
+    spyOnProperty(windowRef, 'nativeWindow').and.returnValue({
+      location: {
+        replace: (val: string) => {}
+      }
+    });
+    const replaceSpy = spyOn(windowRef.nativeWindow.location, 'replace');
+
+    component.initCreationProcess();
+
+    expect(urlServiceSpy).toHaveBeenCalled();
+    expect(replaceSpy).toHaveBeenCalledWith('/creator-dashboard?mode=create');
+  });
+
+  it('should open a create activity modal', () => {
+    component.creationInProgress = false;
+    component.canCreateCollections = true;
+    const urlServiceSpy = spyOn(urlService, 'getPathname').and.returnValue(
+      '/creator-dashboard');
+    const modalSpy = spyOn(ngbModal, 'open').and.callFake((dlg, opt) => {
+      setTimeout(opt.beforeDismiss);
```

Comment:
Why do you need this? In the past timeouts in the code have resulted in flaky tests... Just want to make sure this is needed, before adding it.

Reply:
Yep, we don't need that, moreover, `opt.beforeDismiss` is undefined in this case. Removed setTimeout. 

(We are using this at many other places in the codebase, may need to check them too once)